src/fixate/core/switching.py

```diff
@@ -0,0 +1,720 @@
+            mux_b = ("a1_b0", "a1_b1", "a1_b2", None)
+
+            map_tree = (
+            "a0",
+            mux_b,
+            shift_nested(mux_c, [2]),  # 2 in indicative on how many pins to skip. This case is (x2, x3) from mux_b
+            "a3")
+
+        """
+        signal_map: SignalMap = dict()
+
+        bits_at_this_level = (len(tree) - 1).bit_length()
+        pins_at_this_level = pins[:bits_at_this_level]
+
+        for signal_or_tree, pins_for_signal in zip(tree, generate_bit_sets(pins_at_this_level)):
+            if signal_or_tree is None:
+                continue
+            if isinstance(signal_or_tree, Signal):
+                signal_map[signal_or_tree] = set(pins_for_signal) | fixed_pins
+            else:
+                signal_map.update(self._map_tree(
+                    tree=signal_or_tree,
+                    pins=pins[bits_at_this_level:],
+                    fixed_pins=set(pins_for_signal) | fixed_pins,
+                ))
+
+        return signal_map
+
+    def __repr__(self):
+        return self.__class__.__name__
+
+    @staticmethod
+    def _default_update_pins(
+            pin_updates: PinUpdate,
+            trigger_update: bool = True
+    ) -> None:
+        """
+        Output callback to effect a state change in the mux.
+
+        This is a default implementation which simply prints the planned state change to.
+        stdout. When instantiated as part of a jig driver, this will end up connected
+        to an AddressHandler to do the actual pin changes in hardware.
+
+        In general, this method shouldn't be overridden in a subclass. An alternative
+        can be provided to __init__.
+        """
+        print(pin_updates, trigger_update)
+
+
+class VirtualSwitch(VirtualMux):
+    """
+    A VirtualMux that controls a single pin.
+
+    A virtual switch is a multiplexer with a single pin. The multiplex
+    function can accept either boolean values or the strings 'TRUE' and
+    'FALSE'. The virtual address used to switch can be defined as a list
+    with the single element (as if it were a multiplexer) or by using the
+    shorthand which is to define the pin_name attribute as a string.
+    """
+
+    pin_name: Pin = ""
+    map_tree = ("FALSE", "TRUE")
+
+    def multiplex(self, signal_output: Union[Signal, bool], trigger_update: bool = True):
+        if signal_output is True:
+            signal = "TRUE"
+        elif signal_output is False:
+            signal = "FALSE"
+        else:
+            signal = signal_output
+        super().multiplex(signal, trigger_update=trigger_update)
+
+    def __init__(
+        self,
+        update_pins: Optional[PinUpdateCallback] = None,
+    ):
+        if not self.pin_list:
+            self.pin_list = [self.pin_name]
+        super().__init__(update_pins)
+
+
+class RelayMatrixMux(VirtualMux):
+    clearing_time = 0.01
+
+    def _calculate_pins(self, old_signal: Signal, new_signal: Signal) -> tuple[PinSetState, PinSetState]:
```

Comment:
@jcollins1983 said:
> What's the use case for the old_signal in the _calculate_pins method? It appears to be unused currently.

Reply:
Yeah, I will add a comment for that. If you want to implement make-before-break, I think it is needed. Or at least makes it easier/more obvious. And while the underscore marks it as "private" It is also a method that is intended to be overridden, so I though some level of future proofing was reasonable.

Reply:
Added the following
> old_signal isn't currently used, but it is provided in case a future subclass needs to it calculate the update. In particular, it could be useful for make-before-break behaviour.